### ISSUE TITLE
sometimes players lag out too quickly

### ISSUE THREAD

**Issue by andy5995** [ZetaGlest/zetaglest-source#90]
Sometimes during a lag check, players lag out too quickly. There should be a longer grace period before the server disconnects a player.

One way we are trying to address the problem is allowing players to rejoin a game (#20, #21)

Recently I introduced some experimental code to allow a longer grace period:

https://github.com/ZetaGlest/zetaglest-source/commit/e135baba85d2b2e18ed6f4c753dec065e87b34e3

https://github.com/ZetaGlest/zetaglest-source/blob/e135baba85d2b2e18ed6f4c753dec065e87b34e3/source/glest_game/network/server_interface.cpp#L714-L872

**Comment by Jammyjamjamman** [comment]
I think there's a major problem with coming up with a solution for this. I believe the network communication runs on the main thread. Or at least, in my experience, the main thread appears to freeze if a network communication cannot be established with a player.

The reason this is a problem, is because if we make the game wait an infinite amount of time for someone to re-establish communication (which I think would be the ideal solution), then I believe it will lock the game for everyone. We want to wait a long time, but we want the game to continue for everyone else.

**Comment by andy5995** [comment]
My first game with vibe last night, about 10 minutes into it, everything stopped for like ~7 seconds.. I thought for sure once of us would lag out. But things started moving again normally and we finished.

I don't know if my lag patch helped with that or not.

Is there a way the admin (whoever set up the game) can kick a player from a running game? If not, there should be.

**Comment by Jammyjamjamman** [comment]
> Is there a way the admin (whoever set up the game) can kick a player from a running game? If not, there should be.

No, there isn't. A siight problem with this idea is the headless server is a (virtual) public space, so people aren't given powers as strong  as "blocking"/ "kicking". Most online strategies are like this. There is no reason not to implement this for personal servers, however.

**Comment by andy5995** [comment]
@mathusummut introduced some code to help with this. Needs more testing.

**Comment by andy5995** [comment]
>No, there isn't. A siight problem with this idea is the headless server is a (virtual) public space, so people aren't given powers as strong as "blocking"/ "kicking". Most online strategies are like this. There is no reason not to implement this for personal servers, however.

So it should be enabled for people who click "host" game from the game lobby, or click custom game and check the option "publish" on master server.

Best way I can think of is implementing it is by having a `/kick player` command from the chat console.

What should happen for lagging is this:

The game pauses if someone exceeds the threshold, before he goes OOS. While in the paused state, ZG should see if the player's network is within the threshold. If ZG can't detect whether or not a player is synced while it's paused, then after being paused for 5 or 10 seconds , it should resume for 1 second and check. And that should repeat until the network improves, the player disconnects of his own will, or the server host kicks him.

Rejoining is only good for players who have temporary network glitches. If a player's network or computer can't handle a game of ZG, they'll have the same problem after rejoining.

Summary: Players who get disconnected for temporary network glitches is a release-blocker. What happened the other day while playing with @JoaoMarcosCSilva is that he didn't lag out, he stayed connected because of @mathusummut patch, but he went OOS and the OOS messages kept repeating. We should forget about the OOS caused by join/rejoin for now. If we can get it so players don't get dc'ed from temporary network glitches, that would be great progress itself.

Joining will be a great feature when we can implement it, but remember, one of the reasons it was (and is) in such high-demand is because of the lag disconnects.
